sycl/source/detail/global_handler.cpp

```diff
@@ -0,0 +1,137 @@
+GlobalHandler::~GlobalHandler() = default;
+
+GlobalHandler &GlobalHandler::instance() {
+  if (!SyclGlobalObjectsHandler) {
+    const std::lock_guard<SpinLock> Lock{GlobalWritesAllowed};
+    if (!SyclGlobalObjectsHandler) {
+      SyclGlobalObjectsHandler = new GlobalHandler();
+    }
+  }
+
+  return *SyclGlobalObjectsHandler;
+}
+
+Scheduler &GlobalHandler::getScheduler() {
+  if (!MScheduler) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MScheduler) {
+      MScheduler = std::make_unique<Scheduler>();
+    }
+  }
+  return *MScheduler;
+}
+ProgramManager &GlobalHandler::getProgramManager() {
+  if (!MProgramManager) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MProgramManager) {
+      MProgramManager = std::make_unique<ProgramManager>();
+    }
+  }
+  return *MProgramManager;
+}
+Sync &GlobalHandler::getSync() {
+  if (!MSync) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MSync) {
+      MSync = std::make_unique<Sync>();
+    }
+  }
+  return *MSync;
+}
+std::vector<PlatformImplPtr> &GlobalHandler::getPlatformCache() {
+  if (!MPlatformCache) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MPlatformCache) {
+      MPlatformCache = std::make_unique<std::vector<PlatformImplPtr>>();
+    }
+  }
+  return *MPlatformCache;
+}
+std::mutex &GlobalHandler::getPlatformMapMutex() {
+  if (!MPlatformMapMutex) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MPlatformMapMutex) {
+      MPlatformMapMutex = std::make_unique<std::mutex>();
+    }
+  }
+  return *MPlatformMapMutex;
+}
+std::mutex &GlobalHandler::getFilterMutex() {
+  if (!MFilterMutex) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MFilterMutex) {
+      MFilterMutex = std::make_unique<std::mutex>();
+    }
+  }
+  return *MFilterMutex;
+}
+std::vector<plugin> &GlobalHandler::getPlugins() {
+  if (!MPlugins) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MPlugins) {
+      MPlugins = std::make_unique<std::vector<plugin>>();
+    }
+  }
+  return *MPlugins;
+}
+
+void shutdown() {
+  if (SyclGlobalObjectsHandler) {
+    const std::lock_guard<SpinLock> Lock{ShutdownLock};
+    if (SyclGlobalObjectsHandler) {
+      // Acquire fields lock to make sure no thread creates fields while
+      // object destruction is in flight.
+      const std::lock_guard<SpinLock> Lock{
+          SyclGlobalObjectsHandler->getFieldsLock()};
+      delete SyclGlobalObjectsHandler;
+    }
+  }
+}
+
+#ifdef WIN32
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpReserved) {
+  // Perform actions based on the reason for calling.
+  switch (fdwReason) {
+  case DLL_PROCESS_DETACH:
+    shutdown();
+    break;
+  case DLL_PROCESS_ATTACH:
+  case DLL_THREAD_ATTACH:
+  case DLL_THREAD_DETACH:
+    break;
+  }
+  return TRUE; // Successful DLL_PROCESS_ATTACH.
+}
+#else
+__attribute__((destructor)) static void syclUnload() { shutdown(); }
```

Comment:
@alexbatashev Is this extension fully supported in all Linux targets? Does it not require certain version of glibc or binutils?
@gmlueck I think you said in our recent private talk that you had concerns about this approach?
Which?

Reply:
I have two main concerns about this approach.  First, `__attribute((destructor))` is non-standard, so it reduces the portability of the code.  Second, using this attribute starts an "arms race" with the application because the application can also use this attribute.  If it does, the application can still end up calling DPC++ APIs after the `SyclGlobalObjectsHandler` objects have been destroyed.

By my reading of the documentation, using `__attribute((destructor))` with no priority level (as you do here) does not provide any guarantee that the function will be called after other C++ namespace-scope object destructors.  In order to do that, I think you need to specify a priority number as [documented here](https://gcc.gnu.org/onlinedocs/gcc-4.7.0/gcc/Function-Attributes.html).  However, even if you do this, you will still have the "arms race" I describe above.  Applications can also use the attribute with a priority number.  If you go this route, the DPC++ documentation will need to specify that the DPC++ runtime can only be used until a certain destructor priority.

I think it's important to keep in mind that the only problem this PR solves is that it avoids memory leak reports from tools like valgrind.  (At least, that's the only problem I'm aware of.)  The fact that the memory is leaked is totally unimportant because this memory is needed anyway until just before the process exits.  The only problem is that the **report** is not clean.

One alternative would be to enable destruction of these `SyclGlobalObjectsHandler` objects only if an environment variable is set.  Such a design would be similar to what we have in this PR, except that `shutdown()` would check an environment variable and only call `delete` if it was set.  Users could set the envvar if they wanted to run valgrind and wanted to see a clean leak report.  The documentation of the envvar would say that it is only safe to use if the application does NOT use `__attribute((destructor))` with a priority number below a certain value.  My guess is that most customers won't care if the report shows a small amount of leaked memory in the DPC++ runtime, so they won't bother to set the envvar.  However, the envvar would probably be useful for us when looking for true memory leaks in DPC++.

BTW, we should also test that this solution does, in fact, solve the problem with valgrind.  I don't know how valgrind sets its hook to check for memory leaks.  For example, it might also use the equivalent of `__attribute((destructor))` with a low priority number.  If so, we will need to make sure that our `shutdown()` runs before valgrind's memory leak report.

Another alternative would be to use a custom C++ memory allocator for any non-trivial global objects in the DPC++ runtime.  The custom allocator would allocate from a statically-allocated memory pool (e.g. just a global `char` array).  In addition, these global objects would need to be allocated via `new` and we would never `delete` them.  The custom allocator ensures that their memory is never dynamically allocated, so tools like valgrind won't report a memory leak due to the lack of `delete` calls.  This is a nice solution because it relies only on standard C++ and is therefore completely portable.

The main drawback of this approach is that we need to estimate the total memory size required for all the non-trivial global objects used in the runtime, and then declare a global `char` array with at least that much space.  This becomes particularly hard if one of these objects is a container (e.g. `vector`) with an unknown number of entries.  One way to mitigate this problem is to design our custom allocator so that it "falls back" to regular `malloc` if the static `char` array is exhausted.  If the static pool is too small, the implementation is still perfectly correct, but reports from valgrind will show a memory leak.  If this happens, we can adjust the size of the pool.

Reply:
> custom allocator would allocate from a statically-allocated memory pool

This looks attractive to me except if we really want the destructors of the objects to be called, e.g. if they really have some side-effects important for correct/effective tear-down.

Reply:
> This looks attractive to me except if we really want the destructors of the objects to be called, e.g. if they really have some side-effects important for correct/effective tear-down.

This would only be necessary if the destructor needed to deallocate some resource that won't be automatically deallocated when the process terminates.  There are some system resources like this, but not many.  I don't think we have any destructors in DPC++ that need to deallocate resources like this now.

Reply:
How about dumping something to disk/screen, or hand-shaking between dependent shared libraries? What I want to avoid is that we need this solution anyways, even of most cases can be covered by a cleaner one.

Reply:
These seem speculative.  Do you know of some case like this now?

Reply:
They *are* truly speculative, indeed. And I am in favor of the statically-allocated memory pool.

Reply:
> This would only be necessary if the destructor needed to deallocate some resource that won't be automatically deallocated when the process terminates. 

We cannot leak objects because libsycl.so can be loaded dynamically(dlopen) multiple times during the process lifetime.

Reply:
>  Is this extension fully supported in all Linux targets? Does it not require certain version of glibc or binutils?
It should be, however I have not found any evidence.

Reply:
> By my reading of the documentation, using __attribute((destructor)) with no priority level (as you do here) does not provide any guarantee that the function will be called after other C++ namespace-scope object destructors. In order to do that, I think you need to specify a priority number as documented here. However, even if you do this, you will still have the "arms race" I describe above. Applications can also use the attribute with a priority number. If you go this route, the DPC++ documentation will need to specify that the DPC++ runtime can only be used until a certain destructor priority.

I think we can live with having priority set and note in documentation.

Reply:
There is alternative approach implemented in TBB library which makes global objects reference countable: https://github.com/oneapi-src/oneTBB/blob/427c252e0bb9e191767a62d8a744b21950c343f6/src/tbb/tbb_main.cpp#L68
But if we went this way we would likely need to revisit a lot of places in the sources to avoid changing the reference counter everywhere.

Reply:
I added the highest priority, so that destructor is the last thing run on Linux.